File: tests/Aspire.Hosting.CodeGeneration.TypeScript.Tests/Snapshots/base.verified.ts
// aspire.ts - Core Aspire types: base classes, ReferenceExpression
import { Handle, AspireClient, MarshalledHandle } from './transport.js';

// Re-export transport types for convenience
export { Handle, AspireClient, CapabilityError, registerCallback, unregisterCallback, registerCancellation, unregisterCancellation } from './transport.js';
export type { MarshalledHandle, AtsError, AtsErrorDetails, CallbackFunction } from './transport.js';
export { AtsErrorCodes, isMarshalledHandle, isAtsError, wrapIfHandle } from './transport.js';

// ============================================================================
// Reference Expression
// ============================================================================

/**
 * Represents a reference expression that can be passed to capabilities.
 *
 * Reference expressions are serialized in the protocol as:
 * ```json
 * {
 *   "$expr": {
 *     "format": "redis://{0}:{1}",
 *     "valueProviders": [
 *       { "$handle": "Aspire.Hosting.ApplicationModel/EndpointReference:1" },
 *       { "$handle": "Aspire.Hosting.ApplicationModel/EndpointReference:2" }
 *     ]
 *   }
 * }
 * ```
 *
 * @example
 * ```typescript
 * const redis = await builder.addRedis("cache");
 * const endpoint = await redis.getEndpoint("tcp");
 *
 * // Create a reference expression
 * const expr = refExpr`redis://${endpoint}:6379`;
 *
 * // Use it in an environment variable
 * await api.withEnvironment("REDIS_URL", expr);
 * ```
 */
export class ReferenceExpression {
    private readonly _format: string;
    private readonly _valueProviders: unknown[];

    private constructor(format: string, valueProviders: unknown[]) {
        this._format = format;
        this._valueProviders = valueProviders;
    }

    /**
     * Creates a reference expression from a tagged template literal.
     *
     * @param strings - The template literal string parts
     * @param values - The interpolated values (handles to value providers)
     * @returns A ReferenceExpression instance
     */
    static create(strings: TemplateStringsArray, ...values: unknown[]): ReferenceExpression {
        // Build the format string with {0}, {1}, etc. placeholders
        let format = '';
        for (let i = 0; i < strings.length; i++) {
            format += strings[i];
            if (i < values.length) {
                format += `{${i}}`;
            }
        }

        // Extract handles from values
        const valueProviders = values.map(extractHandleForExpr);

        return new ReferenceExpression(format, valueProviders);
    }

    /**
     * Serializes the reference expression for JSON-RPC transport.
     * Uses the $expr format recognized by the server.
     */
    toJSON(): { $expr: { format: string; valueProviders?: unknown[] } } {
        return {
            $expr: {
                format: this._format,
                valueProviders: this._valueProviders.length > 0 ? this._valueProviders : undefined
            }
        };
    }

    /**
     * String representation for debugging.
     */
    toString(): string {
        return `ReferenceExpression(${this._format})`;
    }
}

/**
 * Extracts a value for use in reference expressions.
 * Supports handles (objects) and string literals.
 * @internal
 */
function extractHandleForExpr(value: unknown): unknown {
    if (value === null || value === undefined) {
        throw new Error('Cannot use null or undefined in reference expression');
    }

    // String literals - include directly in the expression
    if (typeof value === 'string') {
        return value;
    }

    // Number literals - convert to string
    if (typeof value === 'number') {
        return String(value);
    }

    // Handle objects - get their JSON representation
    if (value instanceof Handle) {
        return value.toJSON();
    }

    // Objects with $handle property (already in handle format)
    if (typeof value === 'object' && value !== null && '$handle' in value) {
        return value;
    }

    // Objects with toJSON that returns a handle
    if (typeof value === 'object' && value !== null && 'toJSON' in value && typeof value.toJSON === 'function') {
        const json = value.toJSON();
        if (json && typeof json === 'object' && '$handle' in json) {
            return json;
        }
    }

    throw new Error(
        `Cannot use value of type ${typeof value} in reference expression. ` +
        `Expected a Handle, string, or number.`
    );
}

/**
 * Tagged template function for creating reference expressions.
 *
 * Use this to create dynamic expressions that reference endpoints, parameters, and other
 * value providers. The expression is evaluated at runtime by Aspire.
 *
 * @example
 * ```typescript
 * const redis = await builder.addRedis("cache");
 * const endpoint = await redis.getEndpoint("tcp");
 *
 * // Create a reference expression using the tagged template
 * const expr = refExpr`redis://${endpoint}:6379`;
 *
 * // Use it in an environment variable
 * await api.withEnvironment("REDIS_URL", expr);
 * ```
 */
export function refExpr(strings: TemplateStringsArray, ...values: unknown[]): ReferenceExpression {
    return ReferenceExpression.create(strings, ...values);
}

// ============================================================================
// ResourceBuilderBase
// ============================================================================

/**
 * Base class for resource builders (e.g., RedisBuilder, ContainerBuilder).
 * Provides handle management and JSON serialization.
 */
export class ResourceBuilderBase<THandle extends Handle = Handle> {
    constructor(protected _handle: THandle, protected _client: AspireClient) {}

    toJSON(): MarshalledHandle { return this._handle.toJSON(); }
}

// ============================================================================
// AspireList<T> - Mutable List Wrapper
// ============================================================================

/**
 * Wrapper for a mutable .NET List<T>.
 * Provides array-like methods that invoke capabilities on the underlying collection.
 *
 * @example
 * ```typescript
 * const items = await resource.getItems(); // Returns AspireList<ItemBuilder>
 * const count = await items.count();
 * const first = await items.get(0);
 * await items.add(newItem);
 * ```
 */
export class AspireList<T> {
    constructor(
        private readonly _handle: Handle,
        private readonly _client: AspireClient,
        private readonly _typeId: string
    ) {}

    /**
     * Gets the number of elements in the list.
     */
    async count(): Promise<number> {
        return await this._client.invokeCapability('Aspire.Hosting/List.length', {
            list: this._handle
        }) as number;
    }

    /**
     * Gets the element at the specified index.
     */
    async get(index: number): Promise<T> {
        return await this._client.invokeCapability('Aspire.Hosting/List.get', {
            list: this._handle,
            index
        }) as T;
    }

    /**
     * Adds an element to the end of the list.
     */
    async add(item: T): Promise<void> {
        await this._client.invokeCapability('Aspire.Hosting/List.add', {
            list: this._handle,
            item
        });
    }

    /**
     * Removes the element at the specified index.
     */
    async removeAt(index: number): Promise<void> {
        await this._client.invokeCapability('Aspire.Hosting/List.removeAt', {
            list: this._handle,
            index
        });
    }

    /**
     * Clears all elements from the list.
     */
    async clear(): Promise<void> {
        await this._client.invokeCapability('Aspire.Hosting/List.clear', {
            list: this._handle
        });
    }

    /**
     * Converts the list to an array (creates a copy).
     */
    async toArray(): Promise<T[]> {
        return await this._client.invokeCapability('Aspire.Hosting/List.toArray', {
            list: this._handle
        }) as T[];
    }

    toJSON(): MarshalledHandle { return this._handle.toJSON(); }
}

// ============================================================================
// AspireDict<K, V> - Mutable Dictionary Wrapper
// ============================================================================

/**
 * Wrapper for a mutable .NET Dictionary<K, V>.
 * Provides object-like methods that invoke capabilities on the underlying collection.
 *
 * @example
 * ```typescript
 * const config = await resource.getConfig(); // Returns AspireDict<string, string>
 * const value = await config.get("key");
 * await config.set("key", "value");
 * const hasKey = await config.containsKey("key");
 * ```
 */
export class AspireDict<K, V> {
    private _resolvedHandle?: Handle;
    private _resolvePromise?: Promise<Handle>;

    constructor(
        private readonly _handleOrContext: Handle,
        private readonly _client: AspireClient,
        private readonly _typeId: string,
        private readonly _getterCapabilityId?: string
    ) {
        // If no getter capability, the handle is already the dictionary handle
        if (!_getterCapabilityId) {
            this._resolvedHandle = _handleOrContext;
        }
    }

    /**
     * Ensures we have the actual dictionary handle by calling the getter if needed.
     */
    private async _ensureHandle(): Promise<Handle> {
        if (this._resolvedHandle) {
            return this._resolvedHandle;
        }
        if (this._resolvePromise) {
            return this._resolvePromise;
        }
        // Call the getter capability to get the actual dictionary handle
        this._resolvePromise = (async () => {
            const result = await this._client.invokeCapability(this._getterCapabilityId!, {
                context: this._handleOrContext
            });
            this._resolvedHandle = result as Handle;
            return this._resolvedHandle;
        })();
        return this._resolvePromise;
    }

    /**
     * Gets the number of key-value pairs in the dictionary.
     */
    async count(): Promise<number> {
        const handle = await this._ensureHandle();
        return await this._client.invokeCapability('Aspire.Hosting/Dict.count', {
            dict: handle
        }) as number;
    }

    /**
     * Gets the value associated with the specified key.
     * @throws If the key is not found.
     */
    async get(key: K): Promise<V> {
        const handle = await this._ensureHandle();
        return await this._client.invokeCapability('Aspire.Hosting/Dict.get', {
            dict: handle,
            key
        }) as V;
    }

    /**
     * Sets the value for the specified key.
     */
    async set(key: K, value: V): Promise<void> {
        const handle = await this._ensureHandle();
        await this._client.invokeCapability('Aspire.Hosting/Dict.set', {
            dict: handle,
            key,
            value
        });
    }

    /**
     * Determines whether the dictionary contains the specified key.
     */
    async containsKey(key: K): Promise<boolean> {
        const handle = await this._ensureHandle();
        return await this._client.invokeCapability('Aspire.Hosting/Dict.has', {
            dict: handle,
            key
        }) as boolean;
    }

    /**
     * Removes the value with the specified key.
     * @returns True if the element was removed; false if the key was not found.
     */
    async remove(key: K): Promise<boolean> {
        const handle = await this._ensureHandle();
        return await this._client.invokeCapability('Aspire.Hosting/Dict.remove', {
            dict: handle,
            key
        }) as boolean;
    }

    /**
     * Clears all key-value pairs from the dictionary.
     */
    async clear(): Promise<void> {
        const handle = await this._ensureHandle();
        await this._client.invokeCapability('Aspire.Hosting/Dict.clear', {
            dict: handle
        });
    }

    /**
     * Gets all keys in the dictionary.
     */
    async keys(): Promise<K[]> {
        const handle = await this._ensureHandle();
        return await this._client.invokeCapability('Aspire.Hosting/Dict.keys', {
            dict: handle
        }) as K[];
    }

    /**
     * Gets all values in the dictionary.
     */
    async values(): Promise<V[]> {
        const handle = await this._ensureHandle();
        return await this._client.invokeCapability('Aspire.Hosting/Dict.values', {
            dict: handle
        }) as V[];
    }

    /**
     * Converts the dictionary to a plain object (creates a copy).
     * Only works when K is string.
     */
    async toObject(): Promise<Record<string, V>> {
        const handle = await this._ensureHandle();
        return await this._client.invokeCapability('Aspire.Hosting/Dict.toObject', {
            dict: handle
        }) as Record<string, V>;
    }

    async toJSON(): Promise<MarshalledHandle> {
        const handle = await this._ensureHandle();
        return handle.toJSON();
    }
}

File: tests/Aspire.Hosting.CodeGeneration.Python.Tests/Snapshots/base.verified.py
# base.py - Core Aspire types: base classes, reference expressions, collections
from __future__ import annotations

from enum import Enum
from typing import Any, Dict, Iterable, List

from transport import AspireClient, Handle


class ReferenceExpression:
    """Represents a reference expression passed to capabilities."""

    def __init__(self, format_string: str, value_providers: List[Any]) -> None:
        self._format_string = format_string
        self._value_providers = value_providers

    @staticmethod
    def create(format_string: str, *values: Any) -> "ReferenceExpression":
        value_providers = [_extract_reference_value(value) for value in values]
        return ReferenceExpression(format_string, value_providers)

    def to_json(self) -> Dict[str, Any]:
        payload: Dict[str, Any] = {"format": self._format_string}
        if self._value_providers:
            payload["valueProviders"] = self._value_providers
        return {"$expr": payload}

    def __str__(self) -> str:
        return f"ReferenceExpression({self._format_string})"


def ref_expr(format_string: str, *values: Any) -> ReferenceExpression:
    """Create a reference expression using a format string."""
    return ReferenceExpression.create(format_string, *values)


class HandleWrapperBase:
    """Base wrapper for ATS handle types."""

    def __init__(self, handle: Handle, client: AspireClient) -> None:
        self._handle = handle
        self._client = client

    def to_json(self) -> Dict[str, str]:
        return self._handle.to_json()


class ResourceBuilderBase(HandleWrapperBase):
    """Base class for resource builder wrappers."""


class AspireList(HandleWrapperBase):
    """Wrapper for mutable list handles with lazy handle resolution."""

    def __init__(
        self,
        handle_or_context: Handle,
        client: AspireClient,
        getter_capability_id: str | None = None
    ) -> None:
        super().__init__(handle_or_context, client)
        self._getter_capability_id = getter_capability_id
        self._resolved_handle: Handle | None = None if getter_capability_id else handle_or_context

    def _ensure_handle(self) -> Handle:
        """Lazily resolve the list handle by calling the getter capability."""
        if self._resolved_handle is not None:
            return self._resolved_handle
        if self._getter_capability_id:
            self._resolved_handle = self._client.invoke_capability(
                self._getter_capability_id,
                {"context": self._handle}
            )
        else:
            self._resolved_handle = self._handle
        return self._resolved_handle

    def count(self) -> int:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/List.length",
            {"list": handle}
        )

    def get(self, index: int) -> Any:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/List.get",
            {"list": handle, "index": index}
        )

    def add(self, item: Any) -> None:
        handle = self._ensure_handle()
        self._client.invoke_capability(
            "Aspire.Hosting/List.add",
            {"list": handle, "item": serialize_value(item)}
        )

    def remove_at(self, index: int) -> bool:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/List.removeAt",
            {"list": handle, "index": index}
        )

    def clear(self) -> None:
        handle = self._ensure_handle()
        self._client.invoke_capability(
            "Aspire.Hosting/List.clear",
            {"list": handle}
        )

    def to_list(self) -> List[Any]:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/List.toArray",
            {"list": handle}
        )

    def to_json(self) -> Dict[str, str]:
        if self._resolved_handle is not None:
            return self._resolved_handle.to_json()
        return self._handle.to_json()


class AspireDict(HandleWrapperBase):
    """Wrapper for mutable dictionary handles with lazy handle resolution."""

    def __init__(
        self,
        handle_or_context: Handle,
        client: AspireClient,
        getter_capability_id: str | None = None
    ) -> None:
        super().__init__(handle_or_context, client)
        self._getter_capability_id = getter_capability_id
        self._resolved_handle: Handle | None = None if getter_capability_id else handle_or_context

    def _ensure_handle(self) -> Handle:
        """Lazily resolve the dict handle by calling the getter capability."""
        if self._resolved_handle is not None:
            return self._resolved_handle
        if self._getter_capability_id:
            self._resolved_handle = self._client.invoke_capability(
                self._getter_capability_id,
                {"context": self._handle}
            )
        else:
            self._resolved_handle = self._handle
        return self._resolved_handle

    def count(self) -> int:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/Dict.count",
            {"dict": handle}
        )

    def get(self, key: str) -> Any:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/Dict.get",
            {"dict": handle, "key": key}
        )

    def set(self, key: str, value: Any) -> None:
        handle = self._ensure_handle()
        self._client.invoke_capability(
            "Aspire.Hosting/Dict.set",
            {"dict": handle, "key": key, "value": serialize_value(value)}
        )

    def contains_key(self, key: str) -> bool:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/Dict.has",
            {"dict": handle, "key": key}
        )

    def remove(self, key: str) -> bool:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/Dict.remove",
            {"dict": handle, "key": key}
        )

    def keys(self) -> List[str]:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/Dict.keys",
            {"dict": handle}
        )

    def values(self) -> List[Any]:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/Dict.values",
            {"dict": handle}
        )

    def to_dict(self) -> Dict[str, Any]:
        handle = self._ensure_handle()
        return self._client.invoke_capability(
            "Aspire.Hosting/Dict.toObject",
            {"dict": handle}
        )

    def to_json(self) -> Dict[str, str]:
        if self._resolved_handle is not None:
            return self._resolved_handle.to_json()
        return self._handle.to_json()


def serialize_value(value: Any) -> Any:
    if isinstance(value, ReferenceExpression):
        return value.to_json()

    if isinstance(value, Handle):
        return value.to_json()

    if hasattr(value, "to_json") and callable(value.to_json):
        return value.to_json()

    if hasattr(value, "to_dict") and callable(value.to_dict):
        return {key: serialize_value(val) for key, val in value.to_dict().items()}

    if isinstance(value, Enum):
        return value.value

    if isinstance(value, list):
        return [serialize_value(item) for item in value]

    if isinstance(value, tuple):
        return [serialize_value(item) for item in value]

    if isinstance(value, dict):
        return {key: serialize_value(val) for key, val in value.items()}

    return value


def _extract_reference_value(value: Any) -> Any:
    if value is None:
        raise ValueError("Cannot use None in reference expressions.")

    if isinstance(value, (str, int, float)):
        return value

    if isinstance(value, Handle):
        return value.to_json()

    if hasattr(value, "to_json") and callable(value.to_json):
        return value.to_json()

    raise ValueError(f"Unsupported reference expression value: {type(value).__name__}")
